File: ruby/scrabble-score/scrabble_score.rb
# frozen_string_literal: true

class Scrabble
  LETTER_SCORE = {
    a: 1, e: 1, i: 1, o: 1, u: 1, l: 1, n: 1, r: 1, s: 1, t: 1,
    d: 2, g: 2,
    b: 3, c: 3, m: 3, p: 3,
    f: 4, h: 4, v: 4, w: 4, y: 4,
    k: 5,
    j: 8, x: 8,
    q: 10, z: 10
  }.transform_keys!(&:to_s)

  def self.score(word)
    new(word).score
  end

  private

  def initialize(word)
    @word = (word || '').downcase.chars
    @score = compute_score
  end

  private_constant :LETTER_SCORE

  attr_reader :word

  def compute_score
    word.sum { |letter| letter_score(letter) }
  end

  def letter_score(letter)
    LETTER_SCORE.fetch(letter, 0)
  end

  public

  attr_reader :score
end

File: clojure/scrabble-score/src/scrabble_score.clj
(ns scrabble-score
  (:require [clojure.string :refer [lower-case]]))

(defn score-letter [letter]
  (case (lower-case letter)
    ("a" "e" "i" "l" "n" "o" "r" "s" "t" "u") 1
    ("d" "g") 2
    ("b" "c" "m" "p") 3
    ("f" "h" "v" "w" "y") 4
    "k" 5
    ("j" "x") 8
    ("q" "z") 10
    :else 0))

(defn score-word [word]
  (reduce + (map score-letter word)))
